### PR TITLE
New driver for Brother wordprocessor/typewriter LW-840ic

### PR DESCRIPTION
**Pull request by BartmanAbyss**
split off of [PR#8202](https://github.com/mamedev/mame/pull/8202)

### CODE REVIEW

File: src/mame/brother/lw840.cpp [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+	PORT_START("kbrow.7")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("SM/Layout")             PORT_CODE(KEYCODE_PRTSCR)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("STORNO")                PORT_CODE(KEYCODE_PAUSE)      PORT_CHAR(UCHAR_MAMEKEY(CANCEL))
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_ENTER)      PORT_CHAR(UCHAR_MAMEKEY(ENTER)) //works
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_BACKSPACE)  PORT_CHAR(UCHAR_MAMEKEY(BACKSPACE)) //works
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    //PORT_CODE(KEYCODE_)
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Horz/Vert")             //PORT_CODE(KEYCODE_)
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_SPACE)      PORT_CHAR(UCHAR_MAMEKEY(SPACE))
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_LSHIFT)     PORT_CHAR(UCHAR_MAMEKEY(LSHIFT)) //works
```
Comment (cuavas) [inline src/mame/brother/lw840.cpp:325]: As noted on the other driver, please use `UCHAR_SHIFT_1` for the shift key, add uppercase `PORT_CHAR` for keys in the main keyboard area, and feel free to use UCS4 literals (`U'ü'` etc.) for the accented Latin letters.

Also, what are the “works” comments about?  Does it imply that the keys without them *don’t* work?

**Review by cuavas — CHANGES_REQUESTED** [review]

File: src/mame/brother/lw840.cpp [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+		auto row = keyboard & 0x0f;
+		if(row <= 8)
+			return io_kbrow[row]->read();
+
+		// seems to be able to control power-on self test if not 0xff
+		return 0xffff & mem_mask;
+	}
+	uint16_t keyboard_r(offs_t, uint16_t mem_mask = ~0) {
+		return (keyboard << 8) & mem_mask; 
+	}
```
Comment (cuavas) [inline src/mame/brother/lw840.cpp:149]: As I mentioned before, the `mem_mask` is for lane select.  You don’t need to check it in cases like this where the result doesn’t depend on which lanes are selected.  You only need to check it if different peripherals are selected depending on the selected lanes or something,

File: src/mame/brother/lw840.cpp [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+	auto vram = reinterpret_cast<uint16_t*>(&sram[0x300 / sizeof(uint16_t)]);
+
+	for(auto y = 0; y < 400; y++) {
+		uint32_t* p = &bitmap.pix(y);
+		for(auto x = 0; x < 640; x += 16) {
+			auto gfx = vram[(y * 640 + x) / 16];
+			for(int i = 15; i >= 0; i--)
+				*p++ = palette[BIT(gfx, i)];
+		}
+	}
```
Comment (cuavas) [inline src/mame/brother/lw840.cpp:213]: This won’t work properly on big Endian hosts.  You need to either adapt it to work with bytes, or use `little_endianize_int16` on the data.  You should also at least restrict the rows you update to the range `cliprect.top()` to `cliprect.bottom()`, inclusive.

**Review by cuavas — COMMENTED** [review]
There’s the question of whether the GM82C765B should go in src/devices/machine/upd765.cpp so other things can potentially use it, too.

File: src/mame/brother/lw840.cpp (hunk 3 above)
Comment (cuavas) [reply src/mame/brother/lw840.cpp:213]: Actually, it's already a `uint16_t` pointer – why do you have the `reinterpret_cast`?

Comment (BartmanAbyss) [reply src/mame/brother/lw840.cpp]: my bad ;) forgot after I changed something else

**Comment by BartmanAbyss** [comment]
thanks

**Pull request merged**